### DISCUSSION TITLE
How to Ensure Completion of an Asynchronous Dispose Function

### DISCUSSION THREAD

**Discussion by satoshi-szk** [rrousselGit/riverpod#3089] [Q&A] [answered]
Thanks for this fantastic library. I've been using it extensively, but I have a question that has come up, so I'm posting it here.

In our application, the backend service class manages subprocesses. The dispose function of this service class is responsible for tasks such as terminating subprocesses. It's crucial to call this function before the Flutter application exits to ensure that subprocesses don't remain in a waiting state, leading to issues like port exhaustion and memory leaks.

I was wondering if I could manage this service class with Riverpod, and I wrote the following code:

```dart
@riverpod
BackendService backendService(BackendServiceRef ref) {
  final service = BackendService(ref.read(anotherServiceProvider));
  ref.onDispose(
    () async {
      await service.dispose();
    },
  );
  return service;
}
```

And then I wrote the following test code:

```dart
void main() {
  test("dispose", () {
    final container = ProviderContainer();
    addTearDown(() {
      container.dispose();
    });
    
    container.read(backendServiceProvider).someMethod();
  });
}
```

At this point, it seems that the test is ending without waiting for the execution of the `dispose` function of the service.

Am I doing something wrong, or is this use case, managing such classes with Riverpod, outside the scope of this library? 

Any feedback is greatly appreciated.

**Comment by rrousselGit** [comment, marked as answer]
There's no built-in way to handle waiting for async disposal.

But feel free to make an issue about it

**Reply by satoshi-szk** [reply]
Thanks for the responsiveness @rrousselGit 
I've considered a bit and found a simple workaround.

```dart
void main() {
  test("dispose", () {
    final container = ProviderContainer();
    addTearDown(() async {
      await container.read(backendServiceProvider).dispose();
      container.dispose();  // for synchronous dispose of other services.
    });
    
    container.read(backendServiceProvider).someMethod();
  });
}
```

If the library provides the cleaner way to do this, that sounds the best. But this workaround should suffice in most cases.
Anyway, thanks for your help.
